Implementation of "Restrict cross-module struct initializers to be delegating" (#2)

jrose-apple opened the pull request:
Implementation of [SE-0189](https://github.com/apple/swift-evolution/blob/master/proposals/0189-restrict-cross-module-struct-initializers.md). See the first commit for how this affects the Apple SDK overlays.

rdar://problem/34777878

slavapestov commented on lib/SILOptimizer/Mandatory/DefiniteInitialization.cpp:
```diff
@@ -631,6 +637,21 @@ LifetimeChecker::LifetimeChecker(const DIMemoryObjectInfo &TheMemory,
+  if (TheMemory.isStructInitSelf() &&
+      !Module.getASTContext().isSwiftVersionAtLeast(5)) {
+    SILLocation fnLoc = TheMemory.getFunction().getLocation();
+    if (auto *ctor = fnLoc.getAsASTNode<ConstructorDecl>()) {
```
I'd rather we encode this in the MarkUninitializedInst than add more places in DI where we reach into the AST.

slavapestov commented on lib/SILOptimizer/Mandatory/DefiniteInitialization.cpp:
```diff
@@ -995,6 +1016,41 @@ void LifetimeChecker::handleStoreUse(unsigned UseID) {
+      ty = assign->getSrc()->getType();
+    else if (auto *copyAddr = dyn_cast<CopyAddrInst>(inst))
+      ty = copyAddr->getSrc()->getType().getObjectType();
+    return ty && ty == TheMemory.MemorySILType;
```
You should check if the value being assigned to is TheMemory.MemoryInst, instead of just comparing types

jrose-apple replied:
Oh, of course. I didn't know how to do this and my brain came up with this silly approximation.

slavapestov commented on lib/SILOptimizer/Mandatory/DefiniteInitialization.cpp:
```diff
@@ -995,6 +1016,41 @@ void LifetimeChecker::handleStoreUse(unsigned UseID) {
+    Type selfTy;
+    SILLocation fnLoc = TheMemory.getFunction().getLocation();
+    if (auto *ctor = fnLoc.getAsASTNode<ConstructorDecl>())
+      selfTy = ctor->getImplicitSelfDecl()->getType()->getInOutObjectType();
```
Why isn't this always just TheMemory.getType()?

jrose-apple replied:
This way preserves (some) sugar, which is nice.

jrose-apple replied:
If there's a better way to get at the decl or sugared type I'd be glad to get rid of the ugly reaching up into the function.

slavapestov replied:
How about `selfTy.getAsNominalOrBoundGenericNominal()->getDeclaredInterfaceType()`?

jrose-apple replied:
That doesn't handle constrained extensions (see test case).

slavapestov replied:
I don't think that matters, instead you can use getDeclaredType() to print the type without any generic arguments at all

jrose-apple replied:
Well, I wanted the generic arguments, but if you think it's better without them then that would make things simpler.

slavapestov commented on test/decl/init/cross-module.swift:
```diff
@@ -0,0 +1,174 @@
+// RUN: %empty-directory(%t)
+// RUN: %target-swift-frontend -emit-module -emit-module-path=%t/OtherModule.swiftmodule %S/Inputs/OtherModule.swift
+// RUN: %target-swift-frontend -emit-sil -verify -I %t -swift-version 4 %s > /dev/null
```
Since this is a SIL test, it should go into test/SILOptimizer/

jrose-apple commented:
@slavapestov, @gottesmm, this is my first attempt at putting the pre-Swift-5 check in DI. What would you like me to change here? [Slava beat me to it as I was making this comment.]

To do:
- Check this correctly for inlinable constructors in the same module as well (what `TypeChecker::diagnoseResilientConstructor` is doing but less precisely).
- Treat all cross-module initializers as delegating in Swift 5 mode and when the other module was built with resilience, and then check that those diagnostics are good.

jrose-apple committed:
DI: Warn on non-delegating cross-module struct initializers

...as detected by initializing an individual field without having
initialized the whole object (via `self = value`).

This only applies in pre-Swift-5 mode because the next commit will
treat all cross-module struct initializers as delegating in Swift 5.

jrose-apple committed:
DI: Suggest "self.init()" for extension initializers of C structs

...unless the struct contains a field that cannot be zero-initialized,
such as a non-nullable pointer.

This suggestion is only made for C structs because 'init()' may not be
the right choice for other structs.

jrose-apple committed:
Improve diagnostics for setting a 'let' property in a delegating init

jrose-apple committed:
Treat cross-module struct initializers as delegating in Swift 5

(and when the struct in question is non-fixed-layout, which was
already implemented)

This ensures that these initializers are never fieldwise in Swift 5
mode, which makes it safe for library authors to add new fields.

jrose-apple committed:
DI: Suggest "self.init()" for C structs in Swift 5 mode too

...being careful to only do it once per initializer. Additionally,
/don't/ offer the suggestion if there was already a conditional
assignment to 'self', because that would wipe it out and the user
should think harder.

jrose-apple committed:
[test] Add a cross-module struct initializer test with private fields

jrose-apple committed:
DI: Handle cross-module initializers for empty structs in pre-Swift-5

These also have to delegate to another initializer even though there
are no stored properties to initialize.

jrose-apple commented:
Done! Definitely needs review, though.

@swift-ci Please test compiler performance

jrose-apple commented:
@swift-ci Please test source compatibility

jrose-apple commented:
@swift-ci Please test compiler performance

jrose-apple commented:
@swift-ci Please test compiler performance

jrose-apple commented:
@swift-ci Please test source compatibility

shahmishal commented:
@swift-ci Please test source compatibility

shahmishal commented:
@swift-ci Please test compiler performance

jrose-apple committed:
Simplify diagnoseResilientConstructor all the way away.

Now that struct initializers "just" fall into the delegating case when
they're made inlinable, the only interesting case is class
initializers, which can be checked in a more direct way than what we
were doing before.

jrose-apple commented:
Aargh. That's what I get for trying to edit files while reinstalling Xcode.

jrose-apple commented:
@swift-ci Please test compiler performance

jrose-apple commented:
@swift-ci Please test source compatibility

shahmishal commented:
@swift-ci Please test source compatibility

shahmishal commented:
@swift-ci Please test compiler performance

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
I *think* these are unrelated failures. The perf tests actually *did* seem to finish, too; the results just didn't get posted because some of the projects had trouble. It looks like the full compiler perf tests produce a massive amount of information too, but in most cases there is no change at all from this except in debug builds, where LLVM IR size went up by at most 0.1% and sometimes even went down.

@swift-ci Please test

jrose-apple commented:
@swift-ci Please smoke test compiler performance

shahmishal commented:
@swift-ci build toolchain

shahmishal commented:
Unable to build the toolchain due to

```
/Users/buildnode/jenkins/workspace/swift-PR-toolchain-osx/branch-master/lldb/source/Plugins/ExpressionParser/Swift/SwiftASTManipulator.cpp:1377:21: error: no member named 'mapTypeOutOfContext' in 'swift::TypeBase'
          var_type->mapTypeOutOfContext());
          ~~~~~~~~  ^
1 error generated.

```

xedin commented:
/cc @slavapestov, @shahmishal you’d want a clean build i think

tkremenek commented:
@swift-ci clean test

jrose-apple commented:
@swift-ci Please clean test Linux

jrose-apple commented:
@swift-ci Please clean test Linux

jrose-apple commented:
@swift-ci Please test

tkremenek commented:
@swift-ci clean test

tkremenek commented:
@shahmishal are the tests even running on this one?

shahmishal commented:
Yes, https://ci.swift.org/job/swift-PR-osx/1613/

shahmishal commented:
Linux passed - https://ci.swift.org/job/swift-PR-Linux/1645/

shahmishal commented:
@swift-ci build toolchain

jrose-apple commented:
And the compiler perf test success/failure isn't significant—what's important is that the numbers didn't go up for this PR, which they didn't. Merging!

jpsim commented on stdlib/public/SDK/Foundation/Decimal.swift:
```diff
@@ -327,7 +328,6 @@ extension Decimal {
             
             var i = UInt32(0)
             // this is a bit ugly but it is the closest approximation of the C initializer that can be expressed here.
```
I think this comment can be removed?

jrose-apple replied:
\*shrug\* I thought the loop still counted as ugly, but maybe the comment was just referring to the tuple of zeros. @phausler?
